source/application/tasks/commonjs/index.js

```diff
@@ -1,17 +1,177 @@
-const mkdirp = promisify(mkdirpNodeback);
-import pathExists from 'path-exists';
+import exists from 'path-exists';
 import merge from 'lodash.merge';
-import getPatterns from '../../../library/utilities/get-patterns';
+import fs from 'q-io/fs';
+import {find, difference} from 'lodash';
+
 import resolvePathFormatString from '../../../library/resolve-utilities/resolve-path-format-string';
+import getPatterns from '../../../library/utilities/get-patterns';
+import getPatternMtimes from '../../../library/utilities/get-pattern-mtimes';
 
 const pkg = require(resolve(process.cwd(), 'package.json'));
+const mkdirp = denodeify(mkdirpNodeback);
+const writeFile = denodeify(writeFileNodeback);
+const readFile = denodeify(readFileNodeback);
+const copy = denodeify(ncpNodeback);
+const stat = denodeify(statNodeback);
+const checksumFile = denodeify(checksum.file);
+
+async function getArtifactMtimes(search, patterns) {
+	const types = Object.keys(patterns.formats)
+		.map(key => patterns.formats[key].name);
+
+	const typedFiles = await* [...new Set(types)].map(async type => {
+		const files = await fs.listTree(resolve(search, 'distribution', type));
+		return files.filter(path => extname(path));
+	});
+
+	const artifactPaths = typedFiles
+		.reduce((flattened, files) => [...flattened, ...files], []);
+
+	const artifactMtimes = await* artifactPaths
+		.map(async path => {
+			const artifactId = dirname(relative(resolve(search, 'distribution'), path).split(sep).join('/'));
+			const patternId = artifactId.split(sep).slice(1).join('/');
+			const stats = await stat(path)
+
+			return {
+				id: artifactId,
+				patternId,
+				mtime: stats.mtime
+			};
+		});
+
+	const artifactRegistry = artifactMtimes.reduce((registry, artifact) => {
+		const item = registry[artifact.patternId] || {
+			id: artifact.patternId,
+			artifacts: [],
+			mtimes: [],
+			types: []
+		};
+
+		item.artifacts.push(artifact.id);
+		item.mtimes.push(artifact.mtime);
+		item.types.push(artifact.id.split('/')[0])
+		registry[artifact.patternId] = item;
+		return registry;
+	}, {});
+
+	return Object.values(artifactRegistry).map(item => {
+		const times = item.mtimes.map(time => {
+			return {
+				stamp: time.getTime(),
+				date: time
+			};
+		}).sort((a, b) => a.stamp - b.stamp);
+
+		item.mtime = times[0].date;
+		return item;
+	});
+}
+
+function getPatternsToBuild(artifacts, patterns) {
+	return pattern => {
+		// Find matching pattern artifact
+		const artifact = find(artifacts, {id: pattern.id});
+
+		// If no pattern artifact is found, build it
+		if (!artifact) {
+			return true;
+		}
+
+		// Build if pattern mtime > artifact mtime
+		if (pattern.mtime.getTime() > artifact.mtime.getTime()) {
+			return true;
+		}
+
+		// Get the types in this pattern
+		const types = [...new Set(pattern.files
+			.map(path => extname(path).slice(1))
+			.filter(Boolean)
+			.map(key => patterns.formats[key])
+			.filter(Boolean)
+			.map(config => config.name))];
```

Comment:
I would recommend to not name this parameter config, but format which is what you expect from the map function two lines above. The wordings in this statement are not that easy to get